### PR TITLE
initial group api implementation

### PR DESCRIPTION
**Pull request by butonic**
test api calls using micro:
```
micro call com.owncloud.api.accounts AccountsService.ListAccounts
```

or 

```
micro call com.owncloud.api.accounts AccountsService.CreateAccount '{"account":{"id":"1-2-3-4", "preferred_name":"bob"}}'
```

etc

### CODE REVIEW

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/settings.go  2
- pkg/server/grpc/server.go  1
- pkg/service/v0/groups.go  13
- pkg/service/v0/accounts.go  13
         

Complexity decreasing per file
==============================
+ pkg/service/v0/service.go  -6
         

Clones added
============
- pkg/service/v0/groups.go  5
- pkg/service/v0/accounts.go  7
         

Clones removed
==============
+ pkg/service/v0/service.go  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/settings.go  2
- pkg/service/v0/accounts.go  13
- pkg/server/grpc/server.go  1
- pkg/service/v0/groups.go  13
         

Complexity decreasing per file
==============================
+ pkg/service/v0/service.go  -3
         

Clones added
============
- pkg/service/v0/accounts.go  7
- pkg/service/v0/groups.go  5
         

Clones removed
==============
+ pkg/service/v0/service.go  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/accounts.go  13
- pkg/server/grpc/server.go  1
- pkg/service/v0/settings.go  2
- pkg/service/v0/groups.go  13
         

Complexity decreasing per file
==============================
+ pkg/service/v0/service.go  -3
         

Clones added
============
- pkg/service/v0/accounts.go  7
- pkg/service/v0/groups.go  5
         

Clones removed
==============
+ pkg/service/v0/service.go  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/groups.go  13
- pkg/service/v0/settings.go  2
- pkg/server/grpc/server.go  1
- pkg/service/v0/accounts.go  13
         

Complexity decreasing per file
==============================
+ pkg/service/v0/service.go  -3
         

Clones added
============
- pkg/service/v0/groups.go  5
- pkg/service/v0/accounts.go  7
         

Clones removed
==============
+ pkg/service/v0/service.go  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

File: pkg/service/v0/groups.go [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+	return
+}
+
+// CreateGroup implements the GroupsServiceHandler interface
+func (s Service) CreateGroup(c context.Context, in *proto.CreateGroupRequest, out *proto.Group) (err error) {
+	var id string
+	if in.Group == nil {
+		return fmt.Errorf("group missing")
+	}
+	if in.Group.Id == "" {
```
Comment (kulmann) [inline pkg/service/v0/groups.go:199]: If the incoming group id is NOT empty, we should check if it already exists and throw an error if it does.

File: pkg/service/v0/groups.go [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+		s.log.Error().Err(err).Str("id", id).Str("path", path).Msg("could not remove group")
+		return
+	}
+
+	if err = s.index.Delete(id); err != nil {
+		s.log.Error().Err(err).Str("id", id).Str("path", path).Msg("could not remove group from index")
+		return
+	}
+
+	// TODO delete memberof relationship in users
```
Comment (kulmann) [inline pkg/service/v0/groups.go:246]: 😁

File: pkg/service/v0/groups.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+	}
+	if alreadyRelated == false {
+		g.Members = append(g.Members, a)
+	}
+
+	// check if we need to add the group to the account
+	alreadyRelated = false
+	for i := range a.MemberOf {
+		if a.MemberOf[i].Id == g.Id {
+			alreadyRelated = true
```
Comment (kulmann) [inline pkg/service/v0/groups.go:292]: Add a `break` in this case. No need to iterate through the rest of the groups

File: pkg/proto/v0/accounts.proto [hunk 4]
```diff
@@ -63,39 +63,42 @@ service GroupsService {
     rpc ListGroups(ListGroupsRequest) returns (ListGroupsResponse) {
         // List method maps to HTTP GET
         option (google.api.http) = {
-            get: "/v0/groups"
+            post: "/api/v0/groups/groups-list",
```
[the remark this answers is not in the document]
Comment (butonic) [reply pkg/proto/v0/accounts.proto:66]: hm this falls apart with the AddMember, RemoveMember and ListMembers rpc ...

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/accounts.go  13
- pkg/server/grpc/server.go  1
- pkg/service/v0/settings.go  2
- pkg/service/v0/groups.go  13
         

Complexity decreasing per file
==============================
+ pkg/service/v0/service.go  -3
         

Clones added
============
- pkg/service/v0/accounts.go  7
- pkg/service/v0/groups.go  5
         

Clones removed
==============
+ pkg/service/v0/service.go  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/accounts.go  13
- pkg/service/v0/groups.go  13
- pkg/service/v0/service.go  3
- pkg/service/v0/settings.go  2
- pkg/server/grpc/server.go  1
         

Clones added
============
- pkg/service/v0/accounts.go  7
- pkg/service/v0/groups.go  5
         

Clones removed
==============
+ pkg/service/v0/service.go  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/accounts.go  13
- pkg/service/v0/service.go  3
- pkg/server/grpc/server.go  1
- pkg/service/v0/groups.go  13
- pkg/service/v0/settings.go  2
         

Clones added
============
- pkg/service/v0/accounts.go  16
- pkg/service/v0/groups.go  6
         

Clones removed
==============
+ pkg/service/v0/service.go  -8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/server/grpc/server.go  1
- pkg/service/v0/service.go  3
- pkg/service/v0/accounts.go  13
- pkg/service/v0/groups.go  13
- pkg/service/v0/settings.go  2
         

Clones added
============
- pkg/service/v0/accounts.go  13
- pkg/service/v0/groups.go  14
         

Clones removed
==============
+ pkg/service/v0/service.go  -8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

File: pkg/service/v0/groups.go (hunk 2 above)
Comment (butonic) [reply pkg/service/v0/groups.go]: implemented

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/accounts.go  13
- pkg/service/v0/settings.go  2
- pkg/service/v0/service.go  3
- pkg/server/grpc/server.go  1
- pkg/service/v0/groups.go  13
         

Clones added
============
- pkg/service/v0/accounts.go  13
- pkg/service/v0/groups.go  14
         

Clones removed
==============
+ pkg/service/v0/service.go  -8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/service.go  3
- pkg/service/v0/accounts.go  13
- pkg/service/v0/settings.go  2
- pkg/server/grpc/server.go  1
- pkg/service/v0/groups.go  13
         

Clones added
============
- pkg/service/v0/accounts.go  13
- pkg/service/v0/groups.go  14
         

Clones removed
==============
+ pkg/service/v0/service.go  -8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/service.go  3
- pkg/server/grpc/server.go  1
- pkg/service/v0/groups.go  13
- pkg/service/v0/accounts.go  13
- pkg/service/v0/settings.go  2
         

Clones added
============
- pkg/service/v0/groups.go  14
- pkg/service/v0/accounts.go  13
         

Clones removed
==============
+ pkg/service/v0/service.go  -8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by butonic** [comment]
test fail because of index returning groups and accounts at the same time ... need to split index or add type to document and filter queries accordingly.

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/accounts.go  13
- pkg/service/v0/settings.go  2
- pkg/service/v0/service.go  3
- pkg/service/v0/groups.go  13
- pkg/server/grpc/server.go  1
         

Clones added
============
- pkg/service/v0/accounts.go  13
- pkg/service/v0/groups.go  14
         

Clones removed
==============
+ pkg/service/v0/service.go  -8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/accounts.go  13
- pkg/service/v0/settings.go  2
- pkg/service/v0/service.go  3
- pkg/server/grpc/server.go  1
- pkg/service/v0/groups.go  13
         

Clones added
============
- pkg/service/v0/accounts.go  13
- pkg/service/v0/groups.go  14
         

Clones removed
==============
+ pkg/service/v0/service.go  -8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by butonic** [comment]
blocked by sonarcloud duplicate code check on options.go and protobuf stuff

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/groups.go  13
- pkg/service/v0/settings.go  2
- pkg/server/grpc/server.go  1
- pkg/service/v0/service.go  3
- pkg/service/v0/accounts.go  13
         

Clones added
============
- pkg/service/v0/groups.go  14
- pkg/service/v0/accounts.go  13
         

Clones removed
==============
+ pkg/service/v0/service.go  -8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-accounts/pullRequest?prid=5750204&bid=18824403)

**Comment by butonic** [comment]
tracking the endpoints in https://github.com/owncloud/ocis/issues/377

**Pull request merged**